### PR TITLE
Use requestAnimationFrame for update loop

### PR DESCRIPTION
**Pull request by Jeremy-Gaillard**
Replaced setTimeout by requestAnimationFrame.

Removed calls to scene update (scene.wait()) from managerCommands and c3Dengine and clarified the update mechanism.

### CODE REVIEW

File: src/Core/Commander/ManagerCommands.js
Comment (tbroyer) [inline src/Core/Commander/ManagerCommands.js]: `return false;` is enough

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js
Comment (peppsac) [inline src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js]: Please remove dead code.

File: src/Core/Commander/ManagerCommands.js
Comment (peppsac) [inline src/Core/Commander/ManagerCommands.js]: Is  `.bind(this)` still needed?

**Comment by Jeremy-Gaillard** [comment]
Applied reviewers' suggestions.

Don't merge yet, there is still a problem with the globe loaded event.

**Comment by Jeremy-Gaillard** [comment]
Ready for review.

**Comment by peppsac** [comment]
So I gave this branch a quick test.
2 issues so far : 
- it takes almost 20 sec to get the first complete rendering of the globe. This is a regression compared to master (3 sec)
- high cpu load during that time and for a little longer (enough to turn on the fans on my laptop :) )

**Comment by Jeremy-Gaillard** [comment]
- What do you mean exactly by complete rendering? Are you waiting for the globe loaded event or the first fully textured globe? In the first case, I do not have such a difference.
- Working on it. I just saw that there are far too much rendering pass.

**Comment by peppsac** [comment]
> What do you mean exactly by complete rendering? Are you waiting for the globe loaded event or the first fully textured globe? In the first case, I do not have such a difference.

Yes, the first fully texture globe. Do you observe the same thing as me?

**Comment by Jeremy-Gaillard** [comment]
Yes. I think it's a problem with the request for the higher resolution tiles being completed before the lower resolution ones.

The performance issues seem to come from the rendering/updating happening each time a command is completed. On the master, 16 commands are executed before each new render.

**Comment by Jeremy-Gaillard** [comment]
I pushed a commit where I wait for multiple commands to be executed before updating. The performance should be far better.

Edit: don't bother yet, can't zoom without freezing

**Comment by Jeremy-Gaillard** [comment]
Should be working now. Needs rebasing however.

File: src/Scene/Scene.js
Comment (peppsac) [inline src/Scene/Scene.js]: Indeed `realtimeSceneProcess` isn't clear. `layersUpdate` ?
Probably worth doing it in this PR since you're cleaning the update loop.

File: src/Scene/Scene.js
Comment (peppsac) [inline src/Scene/Scene.js]: Consider fixing the coding style as well (add `{... }` around if/else blocks)

File: src/Scene/Scene.js
Comment (peppsac) [inline src/Scene/Scene.js]: - add documentation for this method
- Maybe `scene.needsUpdate()` is clearer than `scene.notifyChange()`? (or `scene.invalidate()`?)

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js
Comment (peppsac) [inline src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js]: Are the `//!\\ TEMP` comments still needed?
And I think it should be a `this.scene.notifyChange()`

File: src/Core/Commander/ManagerCommands.js
Comment (peppsac) [inline src/Core/Commander/ManagerCommands.js]: `fillCommandPool` is describing _how_ ManagerCommands does its job but fails to describe _what_ this method is supposed to do.
IMHO `managerCommands.update()`, `managerCommands.execCommands()` are better. WDYT?

File: src/Core/Commander/ManagerCommands.js
Comment (peppsac) [inline src/Core/Commander/ManagerCommands.js]: I think @tbroyer advised against this pattern and recommended using `bind(this)`

File: src/Core/Commander/ManagerCommands.js
Comment (peppsac) [inline src/Core/Commander/ManagerCommands.js]: `currentCommandsCount` -> `runningCommandsCount`?

File: src/Core/Commander/ManagerCommands.js
Comment (peppsac) [inline src/Core/Commander/ManagerCommands.js]: `executedCommands` could have the same `Count` suffix.
Or actually I'd prefer them to leave in a sub-object:

```
this.counters.currentCommands--
this.counters.executedCommands++
```

File: src/Core/Commander/ManagerCommands.js
Comment (peppsac) [inline src/Core/Commander/ManagerCommands.js]: Consider adding a short comment explaining what is going on.

File: src/Scene/Scene.js
Comment (peppsac) [inline src/Scene/Scene.js]: What are these boolean doing? I hope we can get rid of one of them because they seem redundant...

File: src/Scene/Scene.js
Comment (peppsac) [inline src/Scene/Scene.js]: Why are you modifying managerCommand internal state here?

Comment (peppsac) [inline src/Scene/Scene.js]: Except if `executedCommands` means `executedCommandsSinceLastRendering`.
But in this case, consider adding a method to ManagerCommands instead (`resetExecutedCommandsCount`)

File: src/Scene/Scene.js
Comment (peppsac) [inline src/Scene/Scene.js]: This big conditional statement, mixing && and ||, needs to be cleaned/simplified.
At the very least, introduce some simple boolean:

``` js
var atLeast16commandsExecutedSinceLastRender = this.managerCommand.executedCommands >= 16;
var timeout5fps = (timestamp - this.lastTimestamp) > 200; // not sure why you want at least 1 command executed here?
```

File: src/Scene/Scene.js
Comment (peppsac) [inline src/Scene/Scene.js]: Rename to `this.lastRendertime` or `this.lastUpdateTime`

**Comment by peppsac** [comment]
I left a few comments, but overall great job on this PR @Jeremy-Gaillard :+1:

File: src/Core/Commander/ManagerCommands.js
Comment (peppsac) [inline src/Core/Commander/ManagerCommands.js]: I'm not sure that hardcoding a command count is a good idea.
I believe the purpose is two-fold:
- limit the number of parallel network fetches (btw, 16 is too much. On Chrome lots of requests are in 'queued' state, see https://developers.google.com/web/tools/chrome-devtools/profile/network-performance/understanding-resource-timing )
- keep a high enough framerate

For the first point maybe we need a global network fetch limit (16?) + a per-host limit (Chrome documentation suggests 6).
For the second case we need a update duration limit (50 ms if we want to keep 20fps for instance).
Last but not least: some commands are not doing network job (it'll be more visible once #149 is merged).

Comment (tbroyer) [inline src/Core/Commander/ManagerCommands.js]: I haven't really followed the discussion (and haven't looked at the patch besides those few lines), but to keep a high framerate, the idea is generally to flip a flag whenever there's a need to update the "rendering", and in a `requestAnimationFrame` callback to re-render based on the flag.

I have no idea what commands are and how they're used in iTowns, so I can't really comment on how things should work wrt queuing and executing them, but things I've seen elsewhere related to similarly-named patterns are to:
- use [`setImmediate`](https://developer.mozilla.org/en/docs/Web/API/Window/setImmediate) or similar to defer processing (see https://www.nczonline.net/blog/2013/07/09/the-case-for-setimmediate/ )
- run as many commands as possible, timing them until you reach 15ms or so, to keep a 60fps, and deferring the rest to a new `setImmediate` (see https://www.nczonline.net/blog/2009/08/11/timed-array-processing-in-javascript/)

`requestAnimationFrame` should only be used for things that directly affect rendering (so rather as a side-effect of commands IIUC), not as a general-purpose task scheduler.

Comment (peppsac) [inline src/Core/Commander/ManagerCommands.js]: `setImmediate` is non-standard so maybe we should prefer a standard based implementation.
This post https://groups.google.com/a/chromium.org/forum/#!msg/blink-dev/Hn3GxRLXmR0/XP9xcY_gBPQJ explains why Chrome didn't implement `setImmediate` but also details several alternatives.
The last `tick` function in his post matches what we need; it's just missing a call to `requestAnimationFrame`

File: src/Scene/Scene.js
Comment (Jeremy-Gaillard) [inline src/Scene/Scene.js]: The two boolean have different meaning: needsUpdate indicates that the scene is not in a stable state (still needs subdivision or textures...), forceUpdate means the scene has to be rendered, even if the scene has not changed.
The scene can need an update but not update for performance reasons: typically, we want to wait for a certain amount of commands to be executed before rendering and updating the scene.

I clarified all this in a new commit.

File: src/Core/Commander/ManagerCommands.js
Comment (tbroyer) [inline src/Core/Commander/ManagerCommands.js]: `requestIdleCallback`? https://twitter.com/jacobrossi/status/667047851415969792 & http://permalink.gmane.org/gmane.org.w3c.webperf/757 (not more standard, but with public support from several vendors: https://www.chromestatus.com/feature/5572795866021888 / https://platform-status.mozilla.org/#request-idle-callback)

FWIW, that `tick` function is the one I described above, and more or less what can be found in Nicholas Zakas' blog entry I linked to ;-)

Comment (Jeremy-Gaillard) [inline src/Core/Commander/ManagerCommands.js]: I'll look at these links and see if I can adapt them to our case.

File: src/Scene/Scene.js
Comment (peppsac) [inline src/Scene/Scene.js]: What `force` really means is `forceRendering` right?

Comment (Jeremy-Gaillard) [inline src/Scene/Scene.js]: Yes, the name has changed in the latest version : https://github.com/iTowns/itowns2/pull/141/files#diff-a6b95dde1d808f9cea55c3fc36fc7f2bR190

**Comment by Jeremy-Gaillard** [comment]
Rebased and ready for review.

File: src/Core/Commander/ManagerCommands.js
Comment (peppsac) [inline src/Core/Commander/ManagerCommands.js]: why did you restore this code?

File: src/Scene/Scene.js
Comment (peppsac) [inline src/Scene/Scene.js]: `processLayers` is misleading. `processQuadtreeLayer` ?

Though both `updateLayers` and `processLayers` are confusing.
Consider adding some documentation/comments (esp. : why do we need 2 functions? why updateLayers is only needed if `forceRender` is true, etc)

File: src/Core/Commander/ManagerCommands.js
Comment (peppsac) [inline src/Core/Commander/ManagerCommands.js]: Still no per-host command limit ? (see my earlier comment on why it's important)

File: src/Core/Commander/ManagerCommands.js
Comment (Jeremy-Gaillard) [inline src/Core/Commander/ManagerCommands.js]: It's not exactly restored. It is part of this commit a1135483f65c3c07f0568434241697eabbe985c3 that fixed some commands being launched for disposed nodes.
With the addition of the early drop function, the bug seems to have disappeared, I'll remove the commit.

File: src/Scene/Scene.js
Comment (Jeremy-Gaillard) [inline src/Scene/Scene.js]: I'll try to find better names.
updateLayers updates the state of the layer for visualisation (tile visibility, sse...)
processLayers creates commands and cleans the quadtree.

File: src/Core/Commander/ManagerCommands.js
Comment (Jeremy-Gaillard) [inline src/Core/Commander/ManagerCommands.js]: Looking at it.

**Comment by Jeremy-Gaillard** [comment]
Added a limit to the number of commands per host.
I still need to add a few comments before being ready to merge.

File: src/Core/Commander/ManagerCommands.js
Comment (Jeremy-Gaillard) [inline src/Core/Commander/ManagerCommands.js]: So, I don't think this mechanism is really suited to our case. If I understand your links correctly, the goal of the tick function is to execute the maximum of atoms in 16ms. However, we don't have a lot to do at each frame, we just create new promises. The only big operation is the tree browsing which can't (easily) be cut into multiple atoms.
However, it may be useful in the future, when multiple quadtrees (or other data structures) will need updating.

File: src/Core/Commander/ManagerCommands.js
Comment (peppsac) [inline src/Core/Commander/ManagerCommands.js]: Consider doing this operation just once when the command is created (in InterfaceCommander maybe?).

File: src/Core/Commander/ManagerCommands.js
Comment (peppsac) [inline src/Core/Commander/ManagerCommands.js]: I'd suggest doing this instead:

``` js
if (commandIsRunnable(command)) {
  executeCommand(command);
} else {
  this.queueAsync.queue(command);
}
```

(no need to queue/dequeue an immediately runnable command).

You can also remove the call to `this.executeCommands()` which become useless with this change.

File: src/Core/Commander/ManagerCommands.js
Comment (peppsac) [inline src/Core/Commander/ManagerCommands.js]: This function with nested functions looks really complex. We should aim to simplify the code.

File: src/Core/Commander/ManagerCommands.js
Comment (peppsac) [inline src/Core/Commander/ManagerCommands.js]: We could have a sorted commands queue per host so we don't have to implement this skipping mechanism.
Or if we believe that having a single queue is useful (for global sorting for instance) then we could consider changing the underlying storage to better support our use case.

File: src/Core/Commander/ManagerCommands.js
Comment (peppsac) [inline src/Core/Commander/ManagerCommands.js]: What's the point of replacing `0` by `undefined`?

File: src/Scene/Scene.js
Comment (peppsac) [inline src/Scene/Scene.js]: While fixing typo you can also replace `var` with `const`

File: src/Core/Commander/ManagerCommands.js
Comment (peppsac) [inline src/Core/Commander/ManagerCommands.js]: Or in `addCommand`.
And at the same time you could init the `this.counters.hostCommands[host]` to remove all the `if(this.counters.hostCommands[host] === undefined)` checks.

**Comment by peppsac** [comment]
I updated this PR based, on PR #176 and #178 .
It also includes rework of error handling which will move to a different PR at some point.
This is not ready for merging yet, but you can already give it a try.

**Comment by peppsac** [comment]
Branch updated. Still not ready for merging (waiting for PR #176 to land first)

**Comment by peppsac** [comment]
Updated the branch - it's not ready for merging yet.

The last commit is a debug commit, just to share with you what I'm using and gather feedback: do you think it would be useful to merge something like this in itowns?

Here's what it looks like:
tools

You can test it using: `DEV_BUILD=true npm start`

**Comment by gchoqueux** [comment]
Some color layers aren't refined when the camera stops to move.
And there are error messages in NodeProcess.js:155
Uncaught (in promise) TypeError: Cannot read property 'childrenLoaded' of null(…)

**Comment by gchoqueux** [comment]
There are error messages in NodeProcess.js:155
_Uncaught (in promise) TypeError: Cannot read property 'childrenLoaded' of null(…)_

**Comment by gchoqueux** [comment]
The event `'globe-loaded'` is broken. It emits before all layers are loaded

File: src/Renderer/LayeredMaterial.js
Comment (gchoqueux) [inline src/Renderer/LayeredMaterial.js]: `var showDebug = true;` replace by false

**Comment by gchoqueux** [comment]
Some time, color layers flick for refinement.
Tiles do goings and comings in their ancestors : [Video 1](https://drive.google.com/file/d/0B5xBYEUN9JvFUWQ1NWtWTnpHdXc/view?usp=sharing) 
This bug appears after large zoom [Video 2](https://drive.google.com/file/d/0B5xBYEUN9JvFWVdWTmg4cENxT00/view?usp=sharing), the tiles doesn't use the good ancestor.

**Comment by peppsac** [comment]
@gchoqueux I fixed some of the issues you reported : 
- `error messages in NodeProcess.js:155`
- `The event 'globe-loaded' is broken`

I do not experiment flickering here though. If you have a way to reproduce the problem I'be interested to learn about it :-)

**Comment by gchoqueux** [comment]
Debug tools needs new PR. 
and The Travis CI build failed with 'Debug tools'

**Comment by peppsac** [comment]
Meh... debug tools are not part of this PR - I just forgot to not push them.
Anyway, do you still reproduce your issues with this version of the code?

**Comment by gchoqueux** [comment]
> debug tools are not part of this PR

ok no problem

> Anyway, do you still reproduce your issues with this version of the code?

I add comment and new video in [`Issue comment`](https://github.com/iTowns/itowns2/pull/141#issuecomment-252026191)

**Comment by peppsac** [comment]
@gchoqueux branch updated.

**Comment by peppsac** [comment]
Updated again.

**Comment by gchoqueux** [comment]
Textures color aren't loaded.
Selected tile have material.nbTextures[Elevation] = 4 ??

capture du 2016-10-27 15-26-07

**Comment by gchoqueux** [comment]
Tile doesn't load, children are builded but not loaded

capture du 2016-10-27 15-44-03

**Comment by peppsac** [comment]
Does it happen each time?

**Comment by gchoqueux** [comment]
no

**Comment by gchoqueux** [comment]
There are performance problems, particularly during loading's globe 
**TimeLine on master** : There aren't many rendering and time loading is 1,2 second (without evaluate scripting)
master

**TimeLine on updateLoop** :  There are a lot of rendering and time loading is 8,5 second (without evaluate scripting)
loop

**Others examples** start loading (start center different)
- Master Loading is 1,2 second
  master2
- updateLoop Loading is 3.4 second
  loop2

**Comment by peppsac** [comment]
We really need to start working on a test framework to prevent regressions (bug or performance)...
Anyway I pushed a new patch that fix performance problems.

File: index.html
Comment (gchoqueux) [inline index.html]: Add a mechanism in the API, if  a new `tileMatrixSetLimits` is needed then it's added automatically. 
And remove this :

``` js
  "20": {}
```

File: src/Core/Commander/ManagerCommands.js
Comment (gchoqueux) [inline src/Core/Commander/ManagerCommands.js]: what is the use of this function?

File: src/Core/Commander/Providers/TileProvider.js
Comment (gchoqueux) [inline src/Core/Commander/Providers/TileProvider.js]: why empty Promise?

File: src/Scene/BrowseTree.js
Comment (gchoqueux) [inline src/Scene/BrowseTree.js:126]: Why refine node if the node isn't displayed?

File: src/Scene/NodeProcess.js
Comment (gchoqueux) [inline src/Scene/NodeProcess.js]: let or var?

File: src/Scene/BrowseTree.js
Comment (gchoqueux) [inline src/Scene/BrowseTree.js]: why you replace `export default BrowseTree;` in BrowseTree.js

**Review by gchoqueux — CHANGES_REQUESTED** [review]

**Comment by gchoqueux** [comment]
this PR could be separated into two : _'requestAnimationFrame'_ and _'rework ProcessNode and browseTree'_

File: index.html
Comment (peppsac) [reply index.html]: This issue is a WMTS one. I'm going to modify `tileInsideLimit` and level instead.
Basically: tileInsideLimit will not test tile.level against zoom.max ; but before fetching a texture we'll clamp the requested level to zoom.max.
WDYT?

File: src/Core/Commander/ManagerCommands.js
Comment (peppsac) [reply src/Core/Commander/ManagerCommands.js]: Now it's useless indeed. I'll remove it.

File: src/Core/Commander/Providers/TileProvider.js
Comment (peppsac) [reply src/Core/Commander/Providers/TileProvider.js]: I will revert this change.

File: src/Scene/BrowseTree.js
Comment (peppsac) [reply src/Scene/BrowseTree.js:126]: I should add a comment. It's meant to make sure that no tile is stuck in the `!loaded` state.

File: src/Scene/BrowseTree.js
Comment (peppsac) [reply src/Scene/BrowseTree.js]: Good catch, I'll revert this too

File: src/Scene/NodeProcess.js
Comment (peppsac) [reply src/Scene/NodeProcess.js]: I should indeed transform `var`s in `let`s

**Comment by peppsac** [comment]
> this PR could be separated into two : 'requestAnimationFrame' and 'rework ProcessNode and browseTree'

I'm not so sure as there are lots of dependencies between both (change in Scene/BrowseTree where driven by changes in requestAnimationFrame and the other way around).

But let's put this issue aside for a while, I'd rather focus on the content of the PR itself :-)

**Review by gchoqueux — APPROVED** [review]

**Pull request merged**
